riscv-sbi-prop.md

```diff
@@ -0,0 +1,290 @@
+physical address.
+
+"priv" is a private data that caller can use to pass information about
+execution context. The SBI implementation must save a copy so that
+caller can reuse while restoring hart from suspend.
+
+Return the appropriate SBI error code.
+
+```
+int sbi_hart_down()
+```
+It powers off the hart and will be used in cpu-hotplug.
+Only individual hart can remove itself from supervisor mode. It can be
+moved to normal state only by sbi_hart_up function.
+
+Return the appropriate SBI error code.
+
+```
+u32 sbi_hart_state(unsigned long hartid)
+```
+Returns the RISCV_POWER_STATE for a specific hartid. This will help make
+kexec like functionality more robust.
+
+```
+void sbi_system_reset(u32 reset_type)
+```
+Reset the entire system.
+
+## Return error code Table:
+Here are the SBI return error codes defined.
+
+| Error Type               | Value  |
+| -------------------------|:------:|
+|  SBI_ERR_SUCCESS         |  0     |
+|  SBI_ERR_FAILURE         | -1     |
+|  SBI_ERR_NOT_SUPPORTED   | -2     |
+|  SBI_ERR_INVALID_PARAM   | -3     |
+|  SBI_ERR_DENIED          | -4     |
+|  SBI_ERR_INVALID_ADDRESS | -5     |
+
+
+## Power State
+A RISC-V core can exist in any of the following power states.
+
+| Power states         | Value  | Description                                  |
+| -------------------  |:------:|:---------------------------------------------|
+| RISCV_HART_ON        | 0      | Powered up & operational.                    |
+| RISCV_HART_STANDBY   | 1      | Powered up but at reduced energy <br> consumption WFI instruction can be used to achieve this state|
+| RISCV_HART_RETENTION | 2      | Deeper low power state. No reset <br> required but higher wakeup latency|
+| RISCV_HART_OFF       | 3      | Powered off. Reset of the core required after power restore|
+
+**_TODO_**:
+
+We probably also need deeper hart power states or system-level power state reporting
+at some point, and non-HART components (caches, memory, etc).  Likely aligned with
+platform specs, once those start to show up.
+
+## Implementation
+Currently, SBI is implemented as a part of BBL. There is a different SBI implementation
+available in coreboot as well.
+
+Alternatively, a separate open BSD/MIT licensed SBI project can be created that
+can be used by anybody to avoid this kind of SBI fragmentation. This project can
+generate both a firmware binary (to be executed directly in M mode) or a static library
+that can be used by different boot loaders. It will also help individual bootloaders
+to either work from M or S mode without a separate SBI implementation.
+
+## Conclusion
+This proposal is far from perfect and absolutely any suggestion is welcome.
+Obviously, there are many other functionalities that can be added to this proposal.
+However, I just wanted to start with something that is an incremental change at
+best to kick off the discussion. The aim here is to initiate a discussion that
+can lead to a robust SBI specification.
```

Comment:
there is not statement of principles with SBI. Is it intended to be like Alpha PALcode, as in the earlier design? is it intended to be a littler operating system, as Luke's requests imply (hence recreating 1970s BIOS?) How big do you want it to be? It seems here like you are building a little operating system. In my view that's not a good idea. In the LinuxBoot project, which we are doing at Google, Facebook, and others, we're trying to reduce the firmware activity to zero for very good security reasons. I view the growth of the SBI in these proposals as going the wrong direction.

A few other thoughts:
We should also acccommodate the changing language landscape: C is on the
way out, and languages like Rust and Go will dominate systems
programming.

The ABI should accommodate modern multiple-value return languages like
Rust and Go, and not be tied to the single return value restrictions
of C. 

Not all callers will be C and they won't always have C structure rules
for alignment and packing; even different C implementations have
different rules, standards notwithstanding. We need to consider that a
call to an SBI is like an RPC call, and we need to define language
independent formats for struct data. We got burned by this on Blue
Gene Q, where the struct format was not even compiler independent.

The set of services provided should be self describing and statically
enumerable via the FDT. We should not require a call to the SBI to
find out what services it provides.

SBI services should be separable, i.e. we should be able to replace
some without replacing the others. There should not be an overarching
version number for the entire set of firmware services on a single
implementation.

The table of services should be presented via the FDT as text, which
is endian- and alignment- independent. The text should be parseable to
a descriptor and an integer to be used as an SBI call number. While it
is desirable to have a small, contiguous range of call numbers, we
should not require it. This parsing need only be done once at boot.

The SBI should not include self-modifying code.

For any SBI service, it should be possible to statically enumerate
other services it depends on.

It should always be possible for a kernel to replace some or all of
the SBI with its own implementation.

Reply:
@rminnich 
There is no doubt that SBI shouldn't be miniOS or too expansive. That's why Palmer suggested that we should start with only base SBI features and ratify them asap. Next, we can look into System/hart power management/vendor SBI extensions (can be worked in parallel by different group of people).
Any extension that need to be added should be only done if that's the only or best approach to do this. It should be avoided if an alternate sane approach exists.

> The set of services provided should be self describing and statically
> enumerable via the FDT. We should not require a call to the SBI to
> find out what services it provides.
> 

Please check the below answer about DT statements.

> SBI services should be separable, i.e. we should be able to replace
> some without replacing the others. There should not be an overarching
> version number for the entire set of firmware services on a single
> implementation.
> 

Not sure what do you mean. All the extensions except the base one will be optional. So firmware may choose not to implement any of that. We may have minor version number increments per extensions depending on if they take longer to ratify. Is that sufficient ?

> The SBI should not include self-modifying code.
> 
> For any SBI service, it should be possible to statically enumerate
> other services it depends on.
> 
> It should always be possible for a kernel to replace some or all of
> the SBI with its own implementation.

ACK. I will put those in a separate section "statement of principles".